### PR TITLE
ARROW-17521: [Python] Add python bindings for NamedTableProvider for Substrait consumer

### PR DESCRIPTION
**Pull request by vibhatha**
This PR includes a basic version to use NamedTable feature in Substrait. The idea is to provide the flexibility to write Python tests with in-memory PyArrow tables.

### CODE REVIEW

**Comment by vibhatha** [comment]
cc @westonpace @richita

File: cpp/src/arrow/engine/substrait/util.cc [hunk 1]
```diff
@@ -126,6 +131,36 @@ Result<std::shared_ptr<RecordBatchReader>> ExecuteSerializedPlan(
...
+
+  compute::ExecContext exec_context(arrow::default_memory_pool(),
+                                    ::arrow::internal::GetCpuThreadPool(), func_registry);
+  ARROW_ASSIGN_OR_RAISE(auto plan, compute::ExecPlan::Make(&exec_context));
+  SubstraitExecutor executor(std::move(plan), exec_context, conversion_options);
+  RETURN_NOT_OK(executor.Init(substrait_buffer, registry));
+  ARROW_ASSIGN_OR_RAISE(auto sink_reader, executor.Execute());
+  // check closing here, not in destructor, to expose error to caller
+  RETURN_NOT_OK(executor.Close());
+  return sink_reader;
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/util.cc:161]: Rather than copy / paste can we call the other method here?

File: cpp/src/arrow/engine/substrait/util.h [hunk 2]
```diff
@@ -20,18 +20,29 @@
...
+using PythonTableProvider =
+    std::function<Result<std::shared_ptr<Table>>(const std::vector<std::string>&)>;
+
 /// \brief Retrieve a RecordBatchReader from a Substrait plan.
 ARROW_ENGINE_EXPORT Result<std::shared_ptr<RecordBatchReader>> ExecuteSerializedPlan(
     const Buffer& substrait_buffer, const ExtensionIdRegistry* registry = NULLPTR,
     compute::FunctionRegistry* func_registry = NULLPTR);
 
+/// \brief Retrieve a RecordBatchReader from a Substrait plan.
+/// Allows to use NamedT
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/util.h:40]: Finish this comment?

File: python/pyarrow/_substrait.pyx [hunk 3]
```diff
@@ -102,3 +103,40 @@ def get_supported_functions():
...
+cdef shared_ptr[CTable] _process_named_table(dict named_args, const std_vector[c_string]& names):
+    cdef c_string c_name
+    py_names = []
+    for i in range(names.size()):
+        c_name = names[i]
+        py_names.append(frombytes(c_name))
+    return pyarrow_unwrap_table(named_args["provider"](py_names))
+
+
+def run_query_with_provider(plan, table_provider):
```
Comment (westonpace) [inline python/pyarrow/_substrait.pyx:117]: Can we make `table_provider` an optional argument for `run_query` rather than adding a new method?  I don't know if we can do it in cython (I think we can for a `def` method) or not but ideally the python API wouldn't need two methods.

**Review by westonpace — CHANGES_REQUESTED** [review]
This looks great and very much in line with what I was hoping.  I just have a few suggestions to remove some of the duplication and ease the api/maintenance.

**Comment by vibhatha** [comment]
@westonpace one doubt is handling multiple names in the provider. Assume we get multiple tables, how should we generate the output?

File: cpp/src/arrow/engine/substrait/util.cc (hunk 1 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/util.cc:164]: Yes of course, this was merely a WIP proto code I just wrote for the sake of getting an insight from @richtia. I will update the code.

File: python/pyarrow/_substrait.pyx (hunk 3 above)
Comment (vibhatha) [reply python/pyarrow/_substrait.pyx]: Of course, this was just a prototype code to show case for test purpose in a short time. I already refactored the code. I think it covers what you are asking for.

File: cpp/src/arrow/engine/substrait/util.cc [hunk 4]
```diff
@@ -107,19 +111,39 @@ class SubstraitExecutor {
...
   // TODO(ARROW-15732)
+  // retrieve input table from table provider
+  ConversionOptions conversion_options;
+  if (table_provider) {
+    NamedTableProvider named_table_provider =
+        [table_provider](
+            const std::vector<std::string>& names) -> Result<compute::Declaration> {
+      if (names.empty()) {
+        return Status::Invalid("names for NamedTable not provided");
+      }
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/util.cc:131]: It wasn't clear to me if a name was required or not so I added https://github.com/substrait-io/substrait/pull/308

If that merges then we are fine returning an invalid status here since this would be an invalid plan.  However, we can probably reject it in `relation_internal.cc` instead so we don't have to repeat this logic for each binding.

File: cpp/src/arrow/engine/substrait/util.h [hunk 5]
```diff
@@ -20,16 +20,22 @@
...
 
 namespace arrow {
 
 namespace engine {
 
+using PythonTableProvider =
+    std::function<Result<std::shared_ptr<Table>>(const std::vector<std::string>&)>;
+
 /// \brief Retrieve a RecordBatchReader from a Substrait plan.
+/// Allows to use NamedTableProvider in Substrait plan execution.
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/util.h:35]:
```suggestion
```

If we are going to document one param we should document all the params

File: python/pyarrow/_substrait.pyx [hunk 6]
```diff
@@ -25,14 +26,84 @@ from pyarrow.includes.libarrow cimport *
...
+    py_names = []
+
+    # provider function could raise an exception
+    try:
+        for i in range(names.size()):
+            c_name = names[i]
+            py_names.append(frombytes(c_name))
+        return pyarrow_unwrap_table(named_args["provider"](py_names))
+    except Exception:
+        return empty_table
```
Comment (westonpace) [inline python/pyarrow/_substrait.pyx:42]: I don't know that an empty table is the correct thing to return here.  Maybe we can return `CResult[shared_ptr[CTable]]` so that we can return an invalid status?  Ideally we would capture the exception message to put into the status too. I'm not sure if we have an example of that.

File: python/pyarrow/_substrait.pyx [hunk 7]
```diff
@@ -25,14 +26,84 @@ from pyarrow.includes.libarrow cimport *
...
+def run_query(plan, table_provider=None):
     """
     Execute a Substrait plan and read the results as a RecordBatchReader.
 
     Parameters
     ----------
     plan : Buffer
         The serialized Substrait plan to execute.
+    table_provider : object (optional)
+        The function determining input table based on names
```
Comment (westonpace) [inline python/pyarrow/_substrait.pyx:54]:
```suggestion
        A function to resolve any NamedTable relation to a table.  The function
        will receive a single argument which will be a list of strings representing the table
        name and should return a pyarrow.Table.
```

Presumably users of this API will be familiar with Substrait so we can reference `NamedTable` directly.

File: python/pyarrow/_substrait.pyx [hunk 8]
```diff
@@ -25,14 +26,84 @@ from pyarrow.includes.libarrow cimport *
...
+    ...                         "names": ["TABLE_NAME_PLACEHOLDER"]
+    ...                 }
+    ...                 }
+    ...             }}
+    ...             ]
+    ...         }
+    ... '''
+    >>> table_name = "t1"
+    >>> query = tobytes(substrait_query.replace(
+    ...             "TABLE_NAME_PLACEHOLDER", table_name))
```
Comment (westonpace) [inline python/pyarrow/_substrait.pyx:103]:
```suggestion
    >>> substrait_query = '''
    ...         {
    ...             "relations": [
    ...             {"rel": {
    ...                 "read": {
    ...                 "base_schema": {
    ...                     "struct": {
    ...                     "types": [
    ...                                 {"i64": {}}
    ...                             ]
    ...                     },
    ...                     "names": [
    ...                             "x"
    ...                             ]
    ...                 },
    ...                 "namedTable": {
    ...                         "names": ["t1"]
    ...                 }
    ...                 }
    ...             }}
    ...             ]
    ...         }
    ... '''
```
There is no need to use the placeholder pattern here.

File: python/pyarrow/tests/test_substrait.py [hunk 9]
```diff
@@ -165,3 +165,172 @@ def test_get_supported_functions():
...
+                "types": [
+                            {"i64": {}}
+                        ]
+                },
+                "names": [
+                        "x"
+                        ]
+            },
+            "namedTable": {
+                    "names": ["TABLE_NAME_PLACEHOLDER"]
```
Comment (westonpace) [inline python/pyarrow/tests/test_substrait.py:200]: No need to use a placeholder.

File: python/pyarrow/tests/test_substrait.py [hunk 10]
```diff
@@ -165,3 +165,172 @@ def test_get_supported_functions():
...
```
Comment (westonpace) [inline python/pyarrow/tests/test_substrait.py:243]: No need to use a placeholder.

File: python/pyarrow/tests/test_substrait.py [hunk 11]
```diff
@@ -165,3 +165,172 @@ def test_get_supported_functions():
...
+    }
+    """
+    query = tobytes(substrait_query)
+    buf = pa._substrait._parse_json_plan(tobytes(query))
+    exec_message = "names for NamedTable not provided"
+    with pytest.raises(ArrowInvalid, match=exec_message):
+        substrait.run_query(buf, table_provider)
+
+
+def test_named_table_no_named_table():
```
Comment (westonpace) [inline python/pyarrow/tests/test_substrait.py:301]: I'm not sure how much value this test is providing.  At the very least, it doesn't seem to be related to named tables at all and is more just testing the C++ consumer's ability to reject invalid plans.

File: cpp/src/arrow/engine/substrait/util.cc (hunk 4 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/util.cc:131]: Yeah that make sense.

File: python/pyarrow/_substrait.pyx (hunk 6 above)
Comment (vibhatha) [reply python/pyarrow/_substrait.pyx]: How about doing a `raise` instead. 

```bash
pyarrow/tests/test_substrait.py::test_named_table_invalid_table_name
  /Users/vibhatha/venv/pyarrow_dev/lib/python3.10/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: 'pyarrow._substrait._process_named_table'
  
  Traceback (most recent call last):
    File "pyarrow/_substrait.pyx", line 39, in pyarrow._substrait._process_named_table
      return pyarrow_unwrap_table(named_args["provider"](py_names))
    File "/Users/vibhatha/github/fork/arrow/python/pyarrow/tests/test_substrait.py", line 223, in table_provider
      raise Exception("Unrecognized table name")
  Exception: Unrecognized table name
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/Users/vibhatha/github/fork/arrow/python/pyarrow/tests/test_substrait.py", line 252, in test_named_table_invalid_table_name
      substrait.run_query(buf, table_provider)
  ValueError: Error occurred in extracting NamedTable : Unrecognized table name
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
```

I want to fix this warning. Not sure if we can handle it here properly.

Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: `raise` looks good to me.  I had forgotten that we were inside SafeCallIntoPython which handles that for us.  Although in that case do we need to even catch the exception?

Comment (vibhatha) [reply python/pyarrow/_substrait.pyx]: I am not quite sure if there is a `SafeCallIntoPython`, aren't we just wraping `cython` function into `std::function`? 
If we don't handle it here, the `pytests` gives the aforementioned warning. What do you think?

**Comment by westonpace** [comment]
> @westonpace one doubt is handling multiple names in the provider. Assume we get multiple tables, how should we generate the output? 

Each call to the function should return a single table.

If multiple names are provided then they form a single combined name.  I'm not really sure why it is structured that way but I think the idea is a that tablenames are often namespaces, for example `["sales", "orders"]` is a single table `sales.orders` and not two different tables.

A provider function is free to know about as many tables as it wants but it can only return one table each call.

File: python/pyarrow/_substrait.pyx (hunk 6 above)
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: BindFunction used SafeCallIntoPython.  I  think what you suggested was good.

Comment (vibhatha) [reply python/pyarrow/_substrait.pyx]: So should we leave as it is?

**Comment by vibhatha** [comment]
@westonpace any more thoughts on this one? Should we add more tests?

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 12]
```diff
@@ -1000,15 +1000,17 @@ TEST(Substrait, GetRecordBatchReader) {
...
+  test_with_registries(
+      [&substrait_json, table_provider](ExtensionIdRegistry* ext_id_reg,
+                                        compute::FunctionRegistry* func_registry) {
+        ASSERT_OK_AND_ASSIGN(auto buf, SerializeJsonPlan(substrait_json));
+        ASSERT_OK_AND_ASSIGN(auto reader, ExecuteSerializedPlan(*buf, table_provider));
+        ASSERT_OK_AND_ASSIGN(auto table, Table::FromRecordBatchReader(reader.get()));
+        // Note: assuming the binary.parquet file contains fixed amount of records
+        // in case of a test failure, re-evalaute the content in the file
+        EXPECT_EQ(table->num_rows(), 12);
+      });
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde_test.cc:1013]:
```suggestion
  test_with_registries(
      [&substrait_json](ExtensionIdRegistry* ext_id_reg,
                                  compute::FunctionRegistry* func_registry) {
        ASSERT_OK_AND_ASSIGN(auto buf, SerializeJsonPlan(substrait_json));
        ASSERT_OK_AND_ASSIGN(auto reader, ExecuteSerializedPlan(*buf, {}));
        ASSERT_OK_AND_ASSIGN(auto table, Table::FromRecordBatchReader(reader.get()));
        // Note: assuming the binary.parquet file contains fixed amount of records
        // in case of a test failure, re-evalaute the content in the file
        EXPECT_EQ(table->num_rows(), 12);
      });
```
Although...I don't think this is the best change.  We should make `table_provider` an optional argument to `ExecuteSerializedPlan`.

File: cpp/src/arrow/engine/substrait/util.cc [hunk 13]
```diff
@@ -107,19 +111,36 @@ class SubstraitExecutor {
...
+  const ConversionOptions& conversion_options_;
 };
 
 }  // namespace
 
 Result<std::shared_ptr<RecordBatchReader>> ExecuteSerializedPlan(
-    const Buffer& substrait_buffer, const ExtensionIdRegistry* extid_registry,
-    compute::FunctionRegistry* func_registry) {
+    const Buffer& substrait_buffer, const PythonTableProvider& table_provider,
+    const ExtensionIdRegistry* registry, compute::FunctionRegistry* func_registry) {
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/util.cc:121]:
```suggestion
    const Buffer& substrait_buffer,
    const ExtensionIdRegistry* registry, compute::FunctionRegistry* func_registry, const ConversionOptions& conversion_options = {}) {
```

I suspect it will only be a matter of time before we want to specify other conversion options.  Also, by defaulting to `{}` we can avoid changing existing tests that are unrelated to the table provider.

File: python/pyarrow/_substrait.pyx (hunk 6 above)
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: Actually, no, not quite.  I'm still not sure why we are catching the exception and rethrowing.  Wouldn't it be better to just not catch it at all?

Comment (vibhatha) [reply python/pyarrow/_substrait.pyx]: Only because of the warning. 🤔

File: cpp/src/arrow/engine/substrait/util.cc (hunk 13 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/util.cc:121]: One thing, if we do this, I think we need to expose `Declaration` to Cython. Because we have to define the `NamedTableProvider` and bind it to the options. Is that okay?

Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/util.cc:121]: Please ignore my previous comment. It is already exposed in `_exec_plan.pyx`

File: cpp/src/arrow/compute/exec/exec_plan.cc [hunk 14]
```diff
@@ -643,6 +643,10 @@ Declaration Declaration::Sequence(std::vector<Declaration> decls) {
   return out;
 }
 
+bool Declaration::IsValid(ExecFactoryRegistry* registry) const {
+  return registry->GetFactory(this->factory_name).ok() && this->options != nullptr;
```
Comment (vibhatha) [inline cpp/src/arrow/compute/exec/exec_plan.cc:647]: @westonpace should we do a full option verification here? I mean validate for all the supported options? 
I am not quite sure if this is an adequate verification. Since it is not sure if doing that kind of a validation is a kind of redundant thing, I left it for discussion.

File: python/pyarrow/_substrait.pyx [hunk 15]
```diff
@@ -17,22 +17,99 @@
...
+
+    py_names = []
+    for i in range(names.size()):
+        c_name = names[i]
+        py_names.append(frombytes(c_name))
+
+    py_table = named_args["provider"](py_names)    
+    c_in_table = pyarrow_unwrap_table(py_table)
+    c_tablesourceopts = make_shared[CTableSourceNodeOptions](
+        c_in_table, 1 << 20)
```
Comment (vibhatha) [inline python/pyarrow/_substrait.pyx:45]: @westonpace I left this number by looking into an existing binding. What is the best to do here?

File: cpp/src/arrow/compute/exec/exec_plan.cc (hunk 14 above)
Comment (vibhatha) [reply cpp/src/arrow/compute/exec/exec_plan.cc:647]: Understood. I agree, without the plan it is not possible to do the validation properly. Should we create a JIRA to address this concretely in future?

File: python/pyarrow/_substrait.pyx (hunk 15 above)
Comment (vibhatha) [reply python/pyarrow/_substrait.pyx:45]: Oh yes, I missed that.

Comment (vibhatha) [reply python/pyarrow/_substrait.pyx:45]: Shall we fix it here too: https://github.com/apache/arrow/blob/99b57e84277f24e8ec1ddadbb11ef8b4f43c8c89/python/pyarrow/_exec_plan.pyx#L95?

Comment (vibhatha) [reply python/pyarrow/_substrait.pyx:45]: Same PR?

File: cpp/src/arrow/compute/exec/exec_plan.cc (hunk 14 above)
Comment (westonpace) [reply cpp/src/arrow/compute/exec/exec_plan.cc:647]: You can but I don't know how high priority it would be.  I think I'd want to see more users of Acero directly (and not just via Substrait or through bindings) before we worry too much about adding in utility validation at the declaration layer.

File: python/pyarrow/_substrait.pyx (hunk 15 above)
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: Sure

File: cpp/src/arrow/compute/exec/exec_plan.cc (hunk 14 above)
Comment (westonpace) [inline cpp/src/arrow/compute/exec/exec_plan.cc:647]:
```suggestion
  return !this->factory_name.empty() && this->options != nullptr;
```

I think the earlier advice I provided was slightly off.  I think, all we need to do here is distinguish between "default-constructed invalid declarations" and "properly constructed declarations".  Actual validation (e.g. hash-map lookup in the factory) doesn't really belong in this method as it should be really fast.

File: python/pyarrow/includes/libarrow_substrait.pxd [hunk 16]
```diff
@@ -22,10 +22,24 @@ from libcpp.vector cimport vector as std_vector
 from pyarrow.includes.common cimport *
 from pyarrow.includes.libarrow cimport *
 
+# ctypedef CResult[shared_ptr[CTable]] named_table_provider(const std_vector[c_string]&)
```
Comment (westonpace) [inline python/pyarrow/includes/libarrow_substrait.pxd:25]:
```suggestion
```

File: python/pyarrow/_substrait.pyx [hunk 17]
```diff
@@ -17,22 +17,98 @@
...
+    ...                         "names": ["t1"]
+    ...                 }
+    ...                 }
+    ...             }}
+    ...             ]
+    ...         }
+    ... '''
+    >>> buf = pa._substrait._parse_json_plan(tobytes(substrait_query))
+    >>> pa.substrait.run_query_with_provider(buf, table_provider)
+    <pyarrow.lib.RecordBatchReader object at 0x100dcaf10>
```
Comment (westonpace) [inline python/pyarrow/_substrait.pyx:111]: Do these examples get verified at build time?  If so I don't think this line `0x100dcaf10` will be very reliable.  If not, I suppose it is ok, but it is odd to see a random pointer value in docs.  We could maybe just omit this line.

**Review by westonpace — CHANGES_REQUESTED** [review]
A few nits and I think this is ready to go.

File: python/pyarrow/_substrait.pyx (hunk 17 above)
Comment (vibhatha) [reply python/pyarrow/_substrait.pyx:111]: I see... let me come up with a better solution.

**Comment by vibhatha** [comment]
cc @westonpace the failing CI is fixed here: https://github.com/apache/arrow/pull/14113

**Review by westonpace — APPROVED** [review]
Thanks again for the addition!

**Comment by jorisvandenbossche** [comment]
This is a nice improvement!

I am wondering one thing: why does the function take a list of names, instead of a single name? 

I understand from https://github.com/apache/arrow/pull/14024#issuecomment-1236472120 that this is because the name can _potentially_ be a list of names indicating some nested schema. 

But for the simple case of single names, having a function that just gets a string and returns a table would be simpler to implement. And for example that would allow to actually implement this easily with a mapping (eg `table_provide={"table1": table1, "table2: table2}.get`)

Would it be an option to call the function with a single string, in case of length 1 list?

**Comment by vibhatha** [comment]
> This is a nice improvement!
> 
> I am wondering one thing: why does the function take a list of names, instead of a single name?
> 
> I understand from [#14024 (comment)](https://github.com/apache/arrow/pull/14024#issuecomment-1236472120) that this is because the name can _potentially_ be a list of names indicating some nested schema.
> 
> But for the simple case of single names, having a function that just gets a string and returns a table would be simpler to implement. And for example that would allow to actually implement this easily with a mapping (eg `table_provide={"table1": table1, "table2: table2}.get`)
> 
> Would it be an option to call the function with a single string, in case of length 1 list?

Good point. Probably we can create a Cython level wrapper for this.
